File: java/src/main/java/io/valkey/javabenchmark/client/AsyncHelper.java
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.valkey.javabenchmark.client;

import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;

/**
 * Synchronous timing helper for all benchmark clients.
 *
 * <p>Provides a unified timing pattern for driver operations. Each operation is
 * executed <b>inline on the calling thread</b> (which is already a long-lived
 * virtual thread owned by {@link io.valkey.javabenchmark.engine.BenchmarkEngine}),
 * timed with {@code System.nanoTime()}, and wrapped in an already-completed
 * {@link CompletableFuture} for interface compatibility.</p>
 *
 * <h3>Why synchronous?</h3>
 * <p>The engine's VT-per-client architecture already provides one virtual thread
 * per client connection. Spawning an additional per-request VT (as the previous
 * implementation did via {@code CompletableFuture.supplyAsync(..., VIRTUAL_EXECUTOR)})
 * was redundant and created ~500K+ unnecessary VT creations/sec, adding significant
 * ForkJoinPool scheduling overhead without any benefit.</p>
 *
 * @author Ilia Kolominsky
 */
public class AsyncHelper {

    /**
     * Execute a blocking operation synchronously with timing.
     *
     * <p>The operation runs on the calling thread (a long-lived virtual thread)
     * and returns an already-completed future with the timed result.</p>
     *
     * @param operation the operation to execute (may block, e.g., {@code client.set().get()})
     * @param <T>       the result type
     * @return a completed future with the timed result
     */
    public static <T> CompletableFuture<TimedResult<T>> timed(Callable<T> operation) {
        long start = System.nanoTime();
        try {
            T result = operation.call();
            long latencyMicros = (System.nanoTime() - start) / 1000;
            return CompletableFuture.completedFuture(TimedResult.of(result, latencyMicros));
        } catch (Exception e) {
            return CompletableFuture.failedFuture(new RuntimeException(e));
        }
    }

    /**
     * Execute a void blocking operation synchronously with timing.
     *
     * <p>The operation runs on the calling thread (a long-lived virtual thread)
     * and returns an already-completed future with the timed result.</p>
     *
     * @param operation the void operation to execute
     * @return a completed future with the timed result (no value)
     */
    public static CompletableFuture<TimedResult<Void>> timedVoid(ThrowingRunnable operation) {
        long start = System.nanoTime();
        try {
            operation.run();
            long latencyMicros = (System.nanoTime() - start) / 1000;
            return CompletableFuture.completedFuture(TimedResult.ofVoid(latencyMicros));
        } catch (Exception e) {
            return CompletableFuture.failedFuture(new RuntimeException(e));
        }
    }

    /**
     * A runnable that can throw checked exceptions.
     */
    @FunctionalInterface
    public interface ThrowingRunnable {
        void run() throws Exception;
    }
}

File: csharp/src/RespBench/Client/AsyncHelper.cs
/*
 * Copyright 2025 the original author or authors.
 */
using System.Diagnostics;

namespace RespBench.Client;

/// <summary>
/// Synchronous timing helper for all benchmark clients.
///
/// Provides a unified timing pattern for driver operations. Each operation is
/// executed inline on the calling thread (which is already a long-lived Task
/// owned by BenchmarkEngine), timed with Stopwatch, and wrapped in an
/// already-completed Task for interface compatibility.
/// </summary>
public static class AsyncHelper
{
    /// <summary>
    /// Execute a blocking operation synchronously with timing.
    /// Returns an already-completed task with the timed result.
    /// </summary>
    public static Task<TimedResult<T>> Timed<T>(Func<T> operation)
    {
        long start = Stopwatch.GetTimestamp();
        try
        {
            T result = operation();
            long latencyMicros = GetElapsedMicros(start);
            return Task.FromResult(TimedResult<T>.Of(result, latencyMicros));
        }
        catch (Exception e)
        {
            return Task.FromException<TimedResult<T>>(new InvalidOperationException(e.Message, e));
        }
    }

    /// <summary>
    /// Execute a void blocking operation synchronously with timing.
    /// Returns an already-completed task with the timed result.
    /// </summary>
    public static Task<TimedResult<object?>> TimedVoid(Action operation)
    {
        long start = Stopwatch.GetTimestamp();
        try
        {
            operation();
            long latencyMicros = GetElapsedMicros(start);
            return Task.FromResult(TimedResult<object?>.OfVoid(latencyMicros));
        }
        catch (Exception e)
        {
            return Task.FromException<TimedResult<object?>>(new InvalidOperationException(e.Message, e));
        }
    }

    /// <summary>
    /// Execute an async operation with timing.
    /// </summary>
    public static async Task<TimedResult<T>> TimedAsync<T>(Func<Task<T>> operation)
    {
        long start = Stopwatch.GetTimestamp();
        try
        {
            T result = await operation().ConfigureAwait(false);
            long latencyMicros = GetElapsedMicros(start);
            return TimedResult<T>.Of(result, latencyMicros);
        }
        catch (Exception e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }

    /// <summary>
    /// Execute an async void operation with timing.
    /// </summary>
    public static async Task<TimedResult<object?>> TimedVoidAsync(Func<Task> operation)
    {
        long start = Stopwatch.GetTimestamp();
        try
        {
            await operation().ConfigureAwait(false);
            long latencyMicros = GetElapsedMicros(start);
            return TimedResult<object?>.OfVoid(latencyMicros);
        }
        catch (Exception e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }

    private static long GetElapsedMicros(long startTimestamp)
    {
        long elapsed = Stopwatch.GetTimestamp() - startTimestamp;
        return elapsed * 1_000_000 / Stopwatch.Frequency;
    }
}
